front/lib/resources/vault_resource.ts

```diff
@@ -0,0 +1,202 @@
+// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unsafe-declaration-merging
+export interface VaultResource extends ReadonlyAttributesType<VaultModel> {}
+// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+export class VaultResource extends BaseResource<VaultModel> {
+  static model: ModelStatic<VaultModel> = VaultModel;
+
+  constructor(model: ModelStatic<VaultModel>, blob: Attributes<VaultModel>) {
+    super(VaultModel, blob);
+  }
+
+  static async makeNew(blob: CreationAttributes<VaultModel>) {
+    const vault = await VaultModel.create(blob);
+
+    return new this(VaultModel, vault.get());
+  }
+
+  get sId(): string {
+    return VaultResource.modelIdToSId({
+      id: this.id,
+      workspaceId: this.workspaceId,
+    });
+  }
+
+  static modelIdToSId({
+    id,
+    workspaceId,
+  }: {
+    id: ModelId;
+    workspaceId: ModelId;
+  }): string {
+    return makeSId("vault", {
+      id,
+      workspaceId,
+    });
+  }
+
+  static async fetchWorkspaceVaults(
+    auth: Authenticator,
+    transaction?: Transaction
+  ): Promise<VaultResource[]> {
+    const owner = auth.getNonNullableWorkspace();
+
+    const vaults = await this.model.findAll({
+      where: {
+        workspaceId: owner.id,
+      },
+      transaction,
+    });
+
+    return vaults.map((vault) => new this(VaultModel, vault.get()));
+  }
+
+  static async fetchWorkspaceSystemVault(
+    auth: Authenticator,
+    transaction?: Transaction
+  ): Promise<VaultResource> {
+    const owner = auth.getNonNullableWorkspace();
+    const vault = await this.model.findOne({
+      where: {
+        workspaceId: owner.id,
+        kind: "system",
+      },
+      transaction,
+    });
+
+    if (!vault) {
+      throw new Error("System vault not found.");
+    }
+
+    return new this(VaultModel, vault.get());
+  }
+
+  static async fetchWorkspaceGlobalVault(
+    auth: Authenticator,
+    transaction?: Transaction
+  ): Promise<VaultResource> {
+    const owner = auth.getNonNullableWorkspace();
+    const vault = await this.model.findOne({
+      where: {
+        workspaceId: owner.id,
+        kind: "global",
+      },
+      transaction,
+    });
+
+    if (!vault) {
+      throw new Error("Global vault not found.");
+    }
+
+    return new this(VaultModel, vault.get());
+  }
+
+  static async fetchById(
+    auth: Authenticator,
+    sId: string
+  ): Promise<VaultResource | null> {
+    const owner = auth.getNonNullableWorkspace();
+
+    const vaultModelId = getResourceIdFromSId(sId);
+    if (!vaultModelId) {
+      return null;
+    }
+
+    const vault = await this.model.findOne({
+      where: {
+        id: vaultModelId,
+        workspaceId: owner.id,
+      },
+    });
+
+    if (!vault) {
+      return null;
+    }
+
+    return new this(VaultModel, vault.get());
+  }
+
+  static async fetchWorkspaceVaultByName(
```

Comment:
Why do we need this one for? Since we don't have an index on `name` we might want to avoid this expensive query.